GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/UpgradeOrchestratorWithIProductUpgraderTests.cs

```diff
@@ -0,0 +1,202 @@
+
+        [SetUp]
+        public void Setup()
+        {
+            this.Tracer = new MockTracer();
+            this.Output = new MockTextWriter();
+            this.MockPreRunChecker = new MockInstallerPrerunChecker(this.Tracer);
+            this.MockPreRunChecker.Reset();
+            this.MockUpgrader = this.DefaultUpgrader();
+            this.orchestrator = new UpgradeOrchestrator(
+                this.MockUpgrader.Object,
+                this.Tracer,
+                this.MockPreRunChecker,
+                input: null,
+                output: this.Output);
+        }
+
+        [TestCase]
+        public void UpgradesWhenUpgradeAvailable()
+        {
+            this.orchestrator.Execute();
+
+            this.MockUpgrader.Verify(upgrader => upgrader.UpgradeAllowed(out It.Ref<string>.IsAny), Times.Once());
+            this.MockUpgrader.Verify(upgrader => upgrader.TryQueryNewestVersion(out It.Ref<System.Version>.IsAny, out It.Ref<string>.IsAny), Times.Once());
+            this.MockUpgrader.Verify(upgrader => upgrader.TryDownloadNewestVersion(out It.Ref<string>.IsAny), Times.Once());
+            this.MockUpgrader.Verify(upgrader => upgrader.TryRunInstaller(It.IsAny<InstallActionWrapper>(), out It.Ref<string>.IsAny), Times.Once());
+            this.MockUpgrader.Verify(upgrader => upgrader.TryCleanup(out It.Ref<string>.IsAny), Times.Once());
+
+            this.orchestrator.ExitCode.ShouldEqual(ReturnCode.Success);
+        }
+
+        [TestCase]
+        public void NewVersionNotAvailable()
+        {
+            this.MockUpgrader.Setup(upgrader => upgrader.TryQueryNewestVersion(out It.Ref<System.Version>.IsAny, out It.Ref<string>.IsAny))
+                .Callback(new TryGetNewerVersionCallback((out System.Version delegateVersion, out string delegateMessage) =>
+                {
+                    delegateVersion = null;
+                    delegateMessage = string.Empty;
+                }))
+                .Returns(true);
+
+            this.orchestrator.Execute();
+
+            this.MockUpgrader.Verify(upgrader => upgrader.UpgradeAllowed(out It.Ref<string>.IsAny), Times.Once());
+            this.MockUpgrader.Verify(upgrader => upgrader.TryQueryNewestVersion(out It.Ref<System.Version>.IsAny, out It.Ref<string>.IsAny), Times.Once());
+            this.MockUpgrader.Verify(upgrader => upgrader.TryDownloadNewestVersion(out It.Ref<string>.IsAny), Times.Never());
+            this.MockUpgrader.Verify(upgrader => upgrader.TryRunInstaller(It.IsAny<InstallActionWrapper>(), out It.Ref<string>.IsAny), Times.Never());
+
+            this.orchestrator.ExitCode.ShouldEqual(ReturnCode.Success);
+        }
+
+        [TestCase]
+        public void GetNewVersionFails()
+        {
+            string errorMessage = "Authentication error.";
+
+            this.MockUpgrader.Setup(upgrader => upgrader.TryQueryNewestVersion(out It.Ref<System.Version>.IsAny, out It.Ref<string>.IsAny))
```

Comment:
What would you think if we put some of the details behind a method with a descriptive name `SetTryQueryNewestVersionReturnValues(...)`.? I like the idea of being able to easily read the test case to get a high level idea of what the test is doing, and I feel the details of setting the Callback are a bit verbose and make it harder to see the intent...

Reply:
Related, I see a lot of code like this being duplicated:

```
this.MockUpgrader.Verify(upgrader => upgrader.UpgradeAllowed(out It.Ref<string>.IsAny), Times.Once());
this.MockUpgrader.Verify(upgrader => upgrader.TryQueryNewestVersion(out It.Ref<System.Version>.IsAny, out It.Ref<string>.IsAny), Times.Once());
this.MockUpgrader.Verify(upgrader => upgrader.TryDownloadNewestVersion(out It.Ref<string>.IsAny), Times.Never());
```

Could all of the common `Verify` calls that are checking if a method has been called be moved to a common `VerifyOrchastratorExecuteBehavior` type method?  Each of the helper methods we expect to be called (or not) could be passed in.

Example:

```
this.VerifyOrchastratorExecuteBehavior(
    callsUpdatedAllowed: true, 
    callsTryQueryNewestVersion: true, 
    callsTryDownloadNewestVersion: false);
```